drivers/spi/spi_mcux_lpspi.c

```diff
@@ -773,87 +790,54 @@ static void spi_mcux_iodev_start(const struct device *dev)
 		break;
 	default:
 		LOG_ERR("Invalid op code %d for submission %p\n", sqe->op, (void *)sqe);
-
-		spi_mcux_iodev_next(dev, true);
-		rtio_iodev_sqe_err(txn_head, -EINVAL);
-		spi_mcux_iodev_complete(dev, 0);
+		spi_mcux_iodev_complete(dev, -EINVAL);
 		return;
 	}
 
 	data->transfer_len = transfer.dataSize;
 
-	k_spinlock_key_t key = spi_spin_lock(dev);
+	k_spinlock_key_t key = spi_spin_lock(rtio_ctx);
 
 	status = LPSPI_MasterTransferNonBlocking(base, &data->handle,
 						 &transfer);
-	spi_spin_unlock(dev, key);
+	spi_spin_unlock(rtio_ctx, key);
 	if (status != kStatus_Success) {
 		LOG_ERR("Transfer could not start");
-		rtio_iodev_sqe_err(txn_head, -EIO);
-	}
-}
-
-static void spi_mcux_iodev_next(const struct device *dev, bool completion)
-{
-	struct spi_mcux_data *data = dev->data;
-
-	k_spinlock_key_t key = spi_spin_lock(dev);
-
-	if (!completion && data->txn_curr != NULL) {
-		spi_spin_unlock(dev, key);
-		return;
-	}
-
-	struct mpsc_node *next = mpsc_pop(&data->io_q);
-
-	if (next != NULL) {
-		struct rtio_iodev_sqe *next_sqe = CONTAINER_OF(next, struct rtio_iodev_sqe, q);
-
-		data->txn_head = next_sqe;
-		data->txn_curr = next_sqe;
-	} else {
-		data->txn_head = NULL;
-		data->txn_curr = NULL;
-	}
-
-	spi_spin_unlock(dev, key);
-
-	if (data->txn_curr != NULL) {
-		struct spi_dt_spec *spi_dt_spec = data->txn_curr->sqe.iodev->data;
-		struct spi_config *spi_cfg = &spi_dt_spec->config;
-
-		spi_mcux_configure(dev, spi_cfg);
-		spi_context_cs_control(&data->ctx, true);
-		spi_mcux_iodev_start(dev);
+		spi_mcux_iodev_complete(dev, -EIO);
 	}
 }
 
 static void spi_mcux_iodev_submit(const struct device *dev,
 				 struct rtio_iodev_sqe *iodev_sqe)
 {
 	struct spi_mcux_data *data = dev->data;
+	struct spi_rtio *rtio_ctx = data->rtio_ctx;
 
-	mpsc_push(&data->io_q, &iodev_sqe->q);
-	spi_mcux_iodev_next(dev, false);
+	if (spi_rtio_submit(rtio_ctx, iodev_sqe)) {
+		spi_mcux_iodev_prepare_start(dev);
+		spi_mcux_iodev_start(dev);
+	}
 }
 
 static void spi_mcux_iodev_complete(const struct device *dev, int status)
```

Comment:
The i2c_rtio_complete does much the same thing and returns true/false from what I recall to inform the driver whether more work should be started. Perhaps spi_rtio_complete could be an equivalent here.

That would reduce the code duplication further.

It's possible that i2c_rtio and spi_rtio do almost the same things entirely and could be converged into bus_rtio, with the blocking call wrappers still being done per bus type (as they each have their own call parameters). That sort of thing can be done after this though.

Reply:
Just a heads-up: I did come up with a spi_rtio_complete(), it's just that it's very limited as there's no handling of spi_context to perform `spi_context_cs_control()` calls.

I can take another look to further optimize it.

Reply:
At a second pass, I find it a bit hard to optimize (without having `spi_rtio` deal with the `spi_context`, or change the API signature to differ from `i2c_rtio`'s). May we explore this further optimization out of scope for this PR?

> It's possible that i2c_rtio and spi_rtio do almost the same things entirely and could be converged into bus_rtio, with the blocking call wrappers still being done per bus type (as they each have their own call parameters). That sort of thing can be done after this though.

BTW - I do agree we'd gain a lot from this bus_rtio refactoring.